README.md

```diff
@@ -72,6 +72,38 @@ if localDC != "" {
 }
 ```
 
+### Shard-aware port
+
+This version of gocql supports a more robust method of establishing connection for each shard by using _shard aware port_ for native transport.
+It greatly reduces time and the number of connections needed to establish a connection per shard in some cases - ex. when many clients connect at once, or when there are non-shard-aware clients connected to the same cluster.
+
+If you are using a custom Dialer and if your nodes expose the shard-aware port, it is highly recommended to update it so that it uses a specific source port when connecting.
+
+- If you are using a custom `net.Dialer`, you can make your dialer honor the source port by wrapping it in a `gocql.ScyllaShardAwareDialer`:
+  ```go
+  oldDialer := net.Dialer{...}
+  clusterConfig.Dialer := &gocql.ScyllaShardAwareDialer{oldDialer}
+  ```
+- If you are using a custom type implementing `gocql.Dialer`, you can get the source port by using the `gocql.ScyllaGetSourcePort` function:
+  ```go
+  func (d *myDialer) DialContext(ctx context.Context, network, addr string) (net.Conn, error) {
+	  sourcePort := gocql.ScyllaGetSourcePort(ctx)
```

Comment:
This hides the source port parameter to the context. Context values should only hold information that is not necessary for correct operation of the code (this basically boils down to information for metrics and tracing).

Why not introduce a new interface instead? That way we could disable the shard aware port at session creation if the dialer does not implement the interface:

```go
type LocalAddressDialer interface {
    // DialWithLocalAddress establishes a new connection but the local address must be set to the provided value.
    // You can use net.Dialer.LocalAddr in your implementation.
    DialWithLocalAddress(ctx context.Context, network, localAddr, remoteAddr string) (net.Conn, error);
}
```

I see `DialerExt` mentioned in some previous comments, but I don't see it in the code now. Why has it been removed?

Also, we could provide some error value for LocalAddressDialer to return so that we can differentiate bind errors from connect errors (and retry with backoff in the latter case).

Reply:
Ah, I missed comment https://github.com/scylladb/gocql/pull/52#issuecomment-696134248 previously.

I still think having an interface is better than putting the port into context. We still can have the current `ScyllaShardAwareDialer` - it could implement the `DialWithLocalAddress` interface and still work as intended (for wrapping net.Dialer).

Reply:
@martin-sucha

> This hides the source port parameter to the context. Context values should only hold information that is not necessary for correct operation of the code (this basically boils down to information for metrics and tracing).
>
> Why not introduce a new interface instead? That way we could disable the shard aware port at session creation if the dialer does not implement the interface:

Technically speaking, the source port parameter is optional. From the perspective of the driver there is no difference if the application did not update their `Dialer` to connect from specified source port, or it's just behind NAT - new connections will be assigned to wrong shards. The driver will fall back to using non-shard-aware port if it realizes that connections are being handled by wrong shards.

I see your point that introducing a `LocalAddressDialer` interface will allow the driver to check if the `Dialer` supports setting source port so that it won't bother trying to connect to shard-aware port if it doesn't - that's why I included `DialerExt` in the first version.

An option would be add an interface which takes a struct with Scylla-specific parameters (so that we don't need to add a new interface each time we add a parameter):

```go
type ScyllaDialer interface {
    DialContextWithParams(ctx context.Context, network, addr string, params *ScyllaDialerParams) (net.Conn, error)
}

type ScyllaDialerParams struct {
    SourcePort uint16
    // other parameters...
}
```

@mmatczuk WDUT?

> Also, we could provide some error value for LocalAddressDialer to return so that we can differentiate bind errors from connect errors (and retry with backoff in the latter case).

We already do differentiate such errors [with this function](https://github.com/scylladb/gocql/pull/52/files#diff-1f3f811ebab7ca16fee0091fdf3b924053596b45797ae19f41f2ae8c7df54ab2R556-R565), used [here](https://github.com/scylladb/gocql/pull/52/files#diff-1f3f811ebab7ca16fee0091fdf3b924053596b45797ae19f41f2ae8c7df54ab2R528). This code assumes that user's `Dialer` will use `net.Dialer.DialContext`, and will work correctly with errors returned from that function.

However, a custom dialer which does not use `net.Dialer.DialContext` might not be able to indicate through an error that a port is already in use. I'll add an error value for that purpose, as you suggested.

Reply:
> An option would be add an interface which takes a struct with Scylla-specific parameters (so that we don't need to add a new interface each time we add a parameter)

Having a struct defeats the purpose of the interface though. If we add a parameter later that the dialer does not know about, we couldn't check whether it implements the old or new behavior.

I think it's safer to introduce yet another interface in the future if the need arises.

> We already do differentiate such errors with this function, used here. This code assumes that user's Dialer will use net.Dialer.DialContext, and will work correctly with errors returned from that function.

Ah, I missed this part, thanks for pointing it out!

> However, a custom dialer which does not use net.Dialer.DialContext might not be able to indicate through an error that a port is already in use. I'll add an error value for that purpose, as you suggested.

Thanks!